AppImage build on Linux can fail when linuxdeploy-plugin-qt picks a Qt 5 qmake instead of Qt 6

### Operating System and Version

Ubuntu/Debian-based Linux

### Distribution

Source build with CMake + Ninja, `APP_DISTRIBUTION=AppImage`

### Description

The current Linux AppImage packaging step can fail on systems where `linuxdeploy-plugin-qt` auto-detects a Qt 5 `qmake` binary even though NotepadNext is built against Qt 6.

This repo is now a Qt 6 project, and `doc/Building.md` currently instructs Linux users to install Qt 6 packages and then run:

```sh
cmake -S . -B build \
    -G Ninja \
    -DCMAKE_BUILD_TYPE=Release \
    -DAPP_DISTRIBUTION=$DISTRIBUTION
cmake --build build --target appimage --parallel
```

When it failed, I got the following message :
```-- Running input plugin: qt -- 

[qt/stdout] Using qmake: /usr/bin/qmake 

[qt/stderr] linuxdeploy-plugin-qt version 1-alpha (git commit ID 511d51d), GitHub actions build 237 built on 2026-03-07 19:21:17 UTC

[qt/stdout] 

[qt/stdout] Using Qt version:  5.15.13  ( 5 ) 

[qt/stdout]   libEGL.so.1  libGLX.so.0  libGLdispatch.so.0  libOpenGL.so.0  libQt6Core.so.6  libQt6Core5Compat.so.6  libQt6DBus.so.6  libQt6Gui.so.6  libQt6Network.so.6  libQt6PrintSupport.so.6  libQt6Widgets.so.6  libX11.so.6  libXau.so.6  libXdmcp.so.6  libb2.so.1  libblkid.so.1  libbrotlicommon.so.1  libbrotlidec.so.1  libbsd.so.0  libbz2.so.1.0  libc.so.6  libcap.so.2  libcom_err.so.2  libcrypto.so.3  libcurl-gnutls.so.4  libdbus-1.so.3  libdouble-conversion.so.3  libduktape.so.207  libexpat.so.1  libffi.so.8  libfontconfig.so.1  libfreetype.so.6  libgcc_s.so.1  libgcrypt.so.20  libgio-2.0.so.0  libglib-2.0.so.0  libgmodule-2.0.so.0  libgmp.so.10  libgnutls.so.30  libgobject-2.0.so.0  libgomp.so.1  libgpg-error.so.0  libgraphite2.so.3  libgssapi_krb5.so.2  libharfbuzz.so.0  libhogweed.so.6  libicudata.so.74  libicui18n.so.74  libicuuc.so.74  libidn2.so.0  libk5crypto.so.3  libkeyutils.so.1  libkrb5.so.3  libkrb5support.so.0  liblber.so.2  libldap.so.2  liblz4.so.1  liblzma.so.5  libm.so.6  libmd.so.0  libmd4c.so.0  libmount.so.1  libnettle.so.8  libnghttp2.so.14  libp11-kit.so.0  libpcre2-16.so.0  libpcre2-8.so.0  libpng16.so.16  libproxy.so.1  libpsl.so.5  libpxbackend-1.0.so  libresolv.so.2  librtmp.so.1  libsasl2.so.2  libselinux.so.1  libssh.so.4  libstdc++.so.6  libsystemd.so.0  libtasn1.so.6  libunistring.so.5  libxcb.so.1  libxkbcommon.so.0  libz.so.1  libzstd.so.1

[qt/stdout] Found Qt modules:  

[qt/stdout] Extra Qt modules:  

[qt/stdout] ERROR: Could not find Qt modules to deploy 
ERROR: Failed to run plugin: qt (exit code: 1) 
ninja: build stopped: subcommand failed.
```
This seems worth fixing in the build scripts because the current documented Linux/AppImage build path can fail on systems that have both Qt 5 and Qt 6 tooling installed.

### Steps to Reproduce

1. Install the Linux dependencies from `doc/Building.md`
2. Have both Qt 5 and Qt 6 `qmake` binaries available, with `/usr/bin/qmake` resolving to Qt 5
3. Run
```sh
cmake -S . -B build \
    -G Ninja \
    -DCMAKE_BUILD_TYPE=Release \
    -DAPP_DISTRIBUTION=$DISTRIBUTION
cmake --build build --target appimage --parallel
```
4. Observe that `linuxdeploy-plugin-qt` reports `Using qmake: /usr/bin/qmake`
5. Observe that it detects Qt 5 and then fails with `Could not find Qt modules to deploy`

### Debug Information

On my system, the relevant part was:
```
[qt/stdout] Using qmake: /usr/bin/qmake
[qt/stdout] Using Qt version:  5.15.13  ( 5 )
```
while the NotepadNext build itself was linking against Qt 6 libraries.

### Additional Details

I already made a fix for this. I'm gonna make a PR next

--- a/cmake/PackagingLinux.cmake
+++ b/cmake/PackagingLinux.cmake
@@ -6,59 +6,45 @@ set(APPDIR_USR ${APPDIR}/usr)
 set(LINUXDEPLOY ${CMAKE_BINARY_DIR}/linuxdeploy-x86_64.AppImage)
 set(LINUXDEPLOY_QT ${CMAKE_BINARY_DIR}/linuxdeploy-plugin-qt-x86_64.AppImage)
 
-function(resolve_qt6_qmake out_var)
-    set(qt6_qmake_candidates "")
+set(APPIMAGE_ENV_VARS
+    LDAI_OUTPUT=NotepadNext-v${PROJECT_VERSION}-x86_64.AppImage
+)
 
-    find_package(Qt6CoreTools QUIET)
-    if(TARGET Qt6::qmake)
-        get_target_property(qt6_qmake_target Qt6::qmake IMPORTED_LOCATION)
-        if(qt6_qmake_target)
-            list(APPEND qt6_qmake_candidates "${qt6_qmake_target}")
-        endif()
-    endif()
-
-    find_program(qmake6_executable
-        NAMES qmake6 qmake-qt6 qmake
-        HINTS
-            /usr/lib/qt6/bin
-            /usr/lib/${CMAKE_LIBRARY_ARCHITECTURE}/qt6/bin
-    )
-    if(qmake6_executable)
-        list(APPEND qt6_qmake_candidates "${qmake6_executable}")
-    endif()
-
-    list(REMOVE_DUPLICATES qt6_qmake_candidates)
-
-    foreach(qmake_candidate IN LISTS qt6_qmake_candidates)
-        execute_process(
-            COMMAND "${qmake_candidate}" -query QT_VERSION
-            OUTPUT_VARIABLE qmake_qt_version
-            OUTPUT_STRIP_TRAILING_WHITESPACE
-            ERROR_QUIET
-            RESULT_VARIABLE qmake_result
-        )
-        if(qmake_result EQUAL 0 AND qmake_qt_version MATCHES "^6\\.")
-            set(${out_var} "${qmake_candidate}" PARENT_SCOPE)
-            return()
-        endif()
-    endforeach()
-
-    set(${out_var} "" PARENT_SCOPE)
-endfunction()
-
-resolve_qt6_qmake(QT6_QMAKE)
-if(QT6_QMAKE)
-    message(STATUS "Using Qt qmake for AppImage packaging: ${QT6_QMAKE}")
-    set(APPIMAGE_ENV_VARS
-        LDAI_OUTPUT=NotepadNext-v${PROJECT_VERSION}-x86_64.AppImage
-        QMAKE=${QT6_QMAKE}
-    )
+if(DEFINED ENV{QMAKE} AND NOT "$ENV{QMAKE}" STREQUAL "")
+    set(APPIMAGE_QMAKE "$ENV{QMAKE}")
 else()
-    message(WARNING "Could not locate a Qt 6 qmake binary for AppImage packaging")
-    set(APPIMAGE_ENV_VARS
-        LDAI_OUTPUT=NotepadNext-v${PROJECT_VERSION}-x86_64.AppImage
+    find_program(APPIMAGE_QMAKE NAMES qmake)
+endif()
+
+if(NOT APPIMAGE_QMAKE)
+    message(FATAL_ERROR
+        "Could not find qmake for AppImage packaging.\n"
+        "Please configure with a Qt 6 qmake, for example:\n"
+        "  QMAKE=$(which qmake6) cmake -S . -B build -DAPP_DISTRIBUTION=AppImage"
     )
 endif()
+
+execute_process(
+    COMMAND "${APPIMAGE_QMAKE}" -query QT_VERSION
+    OUTPUT_VARIABLE APPIMAGE_QT_VERSION
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+    ERROR_QUIET
+    RESULT_VARIABLE APPIMAGE_QMAKE_RESULT
+)
+
+if(NOT APPIMAGE_QMAKE_RESULT EQUAL 0 OR NOT APPIMAGE_QT_VERSION MATCHES "^6\\.")
+    message(FATAL_ERROR
+        "AppImage packaging requires a Qt 6 qmake, but CMake found:\n"
+        "  ${APPIMAGE_QMAKE}\n"
+        "Reported Qt version:\n"
+        "  ${APPIMAGE_QT_VERSION}\n"
+        "Please re-run CMake with a Qt 6 qmake, for example:\n"
+        "  QMAKE=$(which qmake6) cmake -S . -B build -DAPP_DISTRIBUTION=AppImage"
+    )
+endif()
+
+message(STATUS "Using qmake for AppImage packaging: ${APPIMAGE_QMAKE}")
+list(APPEND APPIMAGE_ENV_VARS QMAKE=${APPIMAGE_QMAKE})
 
 install(TARGETS NotepadNext
     RUNTIME DESTINATION bin

--- a/cmake/PackagingLinux.cmake
+++ b/cmake/PackagingLinux.cmake
@@ -6,6 +6,60 @@ set(APPDIR_USR ${APPDIR}/usr)
 set(LINUXDEPLOY ${CMAKE_BINARY_DIR}/linuxdeploy-x86_64.AppImage)
 set(LINUXDEPLOY_QT ${CMAKE_BINARY_DIR}/linuxdeploy-plugin-qt-x86_64.AppImage)
 
+function(resolve_qt6_qmake out_var)
+    set(qt6_qmake_candidates "")
+
+    find_package(Qt6CoreTools QUIET)
+    if(TARGET Qt6::qmake)
+        get_target_property(qt6_qmake_target Qt6::qmake IMPORTED_LOCATION)
+        if(qt6_qmake_target)
+            list(APPEND qt6_qmake_candidates "${qt6_qmake_target}")
+        endif()
+    endif()
+
+    find_program(qmake6_executable
+        NAMES qmake6 qmake-qt6 qmake
+        HINTS
+            /usr/lib/qt6/bin
+            /usr/lib/${CMAKE_LIBRARY_ARCHITECTURE}/qt6/bin
+    )
+    if(qmake6_executable)
+        list(APPEND qt6_qmake_candidates "${qmake6_executable}")
+    endif()
+
+    list(REMOVE_DUPLICATES qt6_qmake_candidates)
+
+    foreach(qmake_candidate IN LISTS qt6_qmake_candidates)
+        execute_process(
+            COMMAND "${qmake_candidate}" -query QT_VERSION
+            OUTPUT_VARIABLE qmake_qt_version
+            OUTPUT_STRIP_TRAILING_WHITESPACE
+            ERROR_QUIET
+            RESULT_VARIABLE qmake_result
+        )
+        if(qmake_result EQUAL 0 AND qmake_qt_version MATCHES "^6\\.")
+            set(${out_var} "${qmake_candidate}" PARENT_SCOPE)
+            return()
+        endif()
+    endforeach()
+
+    set(${out_var} "" PARENT_SCOPE)
+endfunction()
+
+resolve_qt6_qmake(QT6_QMAKE)
+if(QT6_QMAKE)
+    message(STATUS "Using Qt qmake for AppImage packaging: ${QT6_QMAKE}")
+    set(APPIMAGE_ENV_VARS
+        LDAI_OUTPUT=NotepadNext-v${PROJECT_VERSION}-x86_64.AppImage
+        QMAKE=${QT6_QMAKE}
+    )
+else()
+    message(WARNING "Could not locate a Qt 6 qmake binary for AppImage packaging")
+    set(APPIMAGE_ENV_VARS
+        LDAI_OUTPUT=NotepadNext-v${PROJECT_VERSION}-x86_64.AppImage
+    )
+endif()
+
 install(TARGETS NotepadNext
     RUNTIME DESTINATION bin
 )
@@ -40,7 +94,7 @@ add_custom_target(download_linuxdeploy
 
 add_custom_target(appimage
     COMMAND ${CMAKE_COMMAND} -E env
-        LDAI_OUTPUT=NotepadNext-v${PROJECT_VERSION}-x86_64.AppImage
+        ${APPIMAGE_ENV_VARS}
         ${LINUXDEPLOY}
         --appdir ${APPDIR}
         --executable ${APPDIR_USR}/bin/NotepadNext

--- a/cmake/PackagingLinux.cmake
+++ b/cmake/PackagingLinux.cmake
@@ -19,7 +19,7 @@ endif()
 if(NOT APPIMAGE_QMAKE)
     message(FATAL_ERROR
         "Could not find qmake for AppImage packaging.\n"
-        "Please configure with a Qt 6 qmake, for example:\n"
+        "Please install the QT 6 qmake or set the QMAKE variable to a Qt 6 qmake and re-run CMake, for example:\n"
         "  QMAKE=$(which qmake6) cmake -S . -B build -DAPP_DISTRIBUTION=AppImage"
     )
 endif()
@@ -38,7 +38,7 @@ if(NOT APPIMAGE_QMAKE_RESULT EQUAL 0 OR NOT APPIMAGE_QT_VERSION MATCHES "^6\\.")
         "  ${APPIMAGE_QMAKE}\n"
         "Reported Qt version:\n"
         "  ${APPIMAGE_QT_VERSION}\n"
-        "Please re-run CMake with a Qt 6 qmake, for example:\n"
+        "Please set the QMAKE variable to a Qt 6 qmake and re-run CMake, for example:\n"
         "  QMAKE=$(which qmake6) cmake -S . -B build -DAPP_DISTRIBUTION=AppImage"
     )
 endif()

